airflow/contrib/auth/backends/keycloak_auth.py

```diff
@@ -0,0 +1,192 @@
+
+
+class KeyCloakUser(models.User):
+
+    def __init__(self, user):
+        self.user = user
+
+    def is_active(self):
+        '''Required by flask_login'''
+        return True
+
+    def is_authenticated(self):
+        '''Required by flask_login'''
+        return True
+
+    def is_anonymous(self):
+        '''Required by flask_login'''
+        return False
+
+    def get_id(self):
+        '''Returns the current user id as required by flask_login'''
+        return self.user.get_id()
+
+    def data_profiling(self):
+        '''Provides access to data profiling tools'''
+        return True
+
+    def is_superuser(self):
+        '''Access all the things'''
+        return True
+
+
+class AuthenticationError(Exception):
+    pass
+
+
+class GHEAuthBackend(object):
+
+    def __init__(self):
+        self.ghe_host = get_config_param('host')
+        self.login_manager = flask_login.LoginManager()
+        self.login_manager.login_view = 'airflow.login'
+        self.flask_app = None
+        self.ghe_oauth = None
+
+    def ghe_api_route(self, leaf):
+
+        return '/'.join(['http:/',
+                         self.ghe_host,
+                         'auth/realms/brent/protocol/openid-connect'])
+
+    def init_app(self, flask_app):
+        self.flask_app = flask_app
+
+        self.login_manager.init_app(self.flask_app)
+
+        self.ghe_oauth = OAuth(self.flask_app).remote_app(
+            'ghe',
+            consumer_key=get_config_param('client_id'),
+            consumer_secret=get_config_param('client_secret'),
+            # need read:org to get team member list
+            request_token_params={'scope': 'openid'},
+            base_url=self.ghe_host,
+            request_token_url=None,
+            access_token_method='POST',
+            access_token_url=''.join(['http://',
+                                      self.ghe_host,
+                                      '/auth/realms/brent/protocol/openid-connect/token']),
+            authorize_url=''.join(['http://',
+                                   self.ghe_host,
+                                   '/auth/realms/brent/protocol/openid-connect/auth']))
+
+        self.login_manager.user_loader(self.load_user)
+
+        self.flask_app.add_url_rule(get_config_param('oauth_callback_route'),
+                                    'ghe_oauth_callback',
+                                    self.oauth_callback)
+
+    def login(self, request):
+        _log.debug('Redirecting user to GHE login')
+        return self.ghe_oauth.authorize(callback=url_for(
+            'ghe_oauth_callback',
+            _external=True,
+            next=request.args.get('next') or request.referrer or None))
+
+    def get_ghe_user_profile_info(self, ghe_token):
+        #resp = self.ghe_oauth.get(self.ghe_api_route('/userinfo'),token=(ghe_token, ''))
+        #resp=self.ghe_oauth.get('http://10.110.13.147:9090/auth/realms/brent/protocol/openid-connect/userinfo',token=(ghe_token,''))
```

Comment:
Please remove the comments

Reply:
I have create a Jira ticket and simply describe the implementation, and this is the link: https://issues.apache.org/jira/browse/AIRFLOW-1836

the comments have been removed.